File: Exercises/101-109 - Small Problems/medium_1/1000_lights.rb
# You have a bank of switches before you numbered from 1 to 1000. Each switch is connected to exactly one
# light that is initially off. You walk down the row of switches, and turn every one of them on.
# Then, you go back to the beginning and toggle switches 2, 4, 6, and so on. Repeat this for switches 3, 6, 9,
# and so on, and keep going until you have been through 1000 repetitions of this process.

# Write a program that determines which lights are on at the end.

# Example with 5 lights:

# round 1: every light is turned on
# round 2: lights 2 and 4 are now off; 1, 3, 5 are on
# round 3: lights 2, 3, and 4 are now off; 1 and 5 are on
# round 4: lights 2 and 3 are now off; 1, 4, and 5 are on
# round 5: lights 2, 3, and 5 are now off; 1 and 4 are on

# The result is that 2 lights are left on, lights 1 and 4.

# With 10 lights, 3 lights are left on: lights 1, 4, and 9.

def toggle(num_lights)
  switches = Hash.new(0)

  1.upto(num_lights) do |light|
    current = light
    loop do
      switches[current] = switches[current] == 1 ? 0 : 1
      current += light
      break if current > num_lights
    end
  end

  switches.select { |key, value| value == 1 }.keys
end

p toggle(1000)

File: Exercises/210 - Small Problems/interpretive-problem-solving/1000-lights.js
/*
You have a bank of switches before you, numbered from 1 to n. Each switch is connected to exactly one light that is initially off. You walk down the row of switches, and toggle every one of them. You go back to the beginning, and on this second pass you toggle switches 2, 4, 6, and so on. On the third pass, you go back again to the beginning and toggle switches 3, 6, 9, and so on. You repeat this process and keep going until you have been through n repetitions.

Write a program that takes one argument, total number of switches, and returns an array of lights that are on after n repetitions.

Examples

function lightsOn(switches) {
  // ..
}

lightsOn(5);   // [1, 4];

// detailed result of each round for 5 lights
// Round 1: all lights are on
// Round 2: lights 2 and 4 are now off; 1, 3, and 5 are on
// Round 3: lights 2, 3, and 4 are now off; 1 and 5 are on
// Round 4: lights 2 and 3 are now off; 1, 4, and 5 are on
// Round 5: lights 2, 3, and 5 are now off; 1 and 4 are on

lightsOn(100);  // [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
*/


/*
- Parameters
  - Integer
    - Number of switches

- Edge
  - Zero
  - Negative
  - Float
  - String

- Return
  - Array
    - All switches that are currently on.

- Data Structure
  - Object
    - switch_number: boolean

- Algorithm
  - Outer loop from 1 to number of switches.
    - Inner loop through all switches. (1 to number of switches)
      - Toggle the switch at j.
      - Increase j by current iteration(i value from outer loop).
  - Filter the object that has a value of true.
    - Get array of all switches (keys)
    - Filter by checking switch[key] === true
*/

function lightsOn(switches) {
  if (typeof switches !== 'number' || switches <= 0 || !isInteger(switches)) {
    return 'ERROR: Input must be an interger greather than 0.';
  }

  var switchStates = {};
  var i;
  var j;

  for (i = 1; i <= switches; i += 1) {
    for (j = i; j <= switches; j += i) {
      if (switchStates[j] === false || switchStates[j] === undefined) {
        switchStates[j] = true;
      } else {
        switchStates[j] = false;
      }
    }
  }

  return getOnSwitches(switchStates);
}

function isInteger(num) {
  return num % 1 === 0;
}

function getOnSwitches(switchStates) {
  var switches = Object.keys(switchStates);

  return switches.filter(function(switchNum) {
    return switchStates[switchNum] === true;
  }).map(Number);
}

// Test Cases
console.log(lightsOn(5));     // [1, 4];
console.log(lightsOn(100));   // [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
console.log(lightsOn(0));     // ERROR: Input must be an interger greather than 0.
console.log(lightsOn(-1));    // ERROR: Input must be an interger greather than 0.
console.log(lightsOn('HI'));  // ERROR: Input must be an interger greather than 0.
console.log(lightsOn(3.14));  // ERROR: Input must be an interger greather than 0.